packages/yew-router/src/service.rs

```diff
@@ -1,90 +1,160 @@
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+enum RouterAction<T: Routable> {
+    Push(T),
+    Replace(T),
+}
+
+/// Specializes the history API for a particular `Routable` type
+pub(crate) struct RouterState<T: Routable> {
+    last_route: Rc<T>,
+    subscribers: Vec<Callback<Rc<T>>>,
+    _listener: HistoryListener,
+}
 
-    Ok(())
+thread_local! {
+    static ROUTER_STATE: RefCell<AnyMap> = RefCell::new(AnyMap::new());
 }
 
-fn push_impl(url: String) {
-    let history = yew::utils::window().history().expect("no history");
-    let base = base_url();
-    let path = match base {
-        Some(base) => {
-            let path = format!("{}{}", base, url);
-            if path.is_empty() {
-                "/".to_string()
+impl<T: Routable> RouterState<T> {
+    fn with<R>(f: impl FnOnce(Entry<Self>) -> R) -> R {
+        ROUTER_STATE.with(|state| f(state.borrow_mut().entry()))
+    }
+    fn try_with<R>(mut f: impl FnMut(Entry<Self>) -> Result<R, T>) -> R {
+        match Self::with(&mut f) {
+            Ok(res) => return res,
+            Err(error) => Self::handle_error(error),
+        };
+        match Self::with(&mut f) {
+            Ok(res) => res,
+            Err(_) => panic!("Bug in `Routable` implementation"),
+        }
+    }
+    fn determine_current_route() -> Result<T, T> {
+        T::from_route(&*history::current())
+    }
+    fn new() -> Result<Self, T> {
+        let last_route = Rc::new(Self::determine_current_route()?);
+        Ok(Self {
+            last_route,
+            subscribers: Vec::new(),
+            _listener: history::register(Callback::from(Self::update)),
+        })
+    }
+
+    fn try_insert(entry: Entry<Self>) -> Result<&mut Self, T> {
+        Ok(match entry {
+            Entry::Occupied(occ) => occ.into_mut(),
+            Entry::Vacant(vac) => vac.insert(Self::new()?),
+        })
+    }
+
+    // We sometimes return a function to run when the state is not borrowed.
+    // This is so that callbacks don't panic if they try to access the state.
+    fn update_inner(
+        entry: Entry<Self>,
+        route: Rc<Route>,
+    ) -> Result<Option<impl FnOnce() + 'static>, T> {
+        let mut occ = match entry {
+            Entry::Occupied(occ) => occ,
+            Entry::Vacant(_) => return Ok(None),
+        };
+        let state = occ.get_mut();
+        Ok(if state.subscribers.is_empty() {
+            occ.remove();
+            None
+        } else {
+            let route = Rc::new(T::from_route(&*route)?);
+            if state.last_route != route {
+                state.last_route = route.clone();
+                let subscribers = state.subscribers.clone();
+                Some(move || {
+                    for subscriber in subscribers {
+                        subscriber.emit(route.clone());
+                    }
+                })
             } else {
-                path
+                None
+            }
+        })
+    }
+
+    fn handle_error(error: T) {
```

Comment:
Nit: this really isn't an error so should we rename this function something else? Something like `handle_not_found`, `handle_404_error` or `handle_404`?

Reply:
👍